razor-fol/src/transform/relationalize.rs

```diff
@@ -0,0 +1,865 @@
+///
+/// [`expand_eqaulity`]: ./struct.EqualityExpander.html#method.expand_equality
+pub struct EqualityExpander {
+    // Is the symbol used to convert equality to a predicate.
+    equality_symbol: String,
+
+    // Is the [`Generator`] instance used to generate new variable names when variables
+    // appear in more than one position.
+    equality_generator: Generator,
+}
+
+impl EqualityExpander {
+    /// Creates a new `EqualityExpander` instance with default generators and symbols:
+    /// * The equality symbol is `=`.
+    /// * Variables appearing in more than one position are distinguished with `~` as the
+    /// prefix, followed by `:` and a unique index.
+    pub fn new() -> Self {
+        Self {
+            equality_symbol: "=".into(),
+            equality_generator: Generator::new().set_prefix("~").set_delimiter(":"),
+        }
+    }
+
+    /// Use `symbol` for equality predicates.
+    pub fn set_equality_symbol<S: Into<String>>(&mut self, symbol: S) {
+        self.equality_symbol = symbol.into();
+    }
+
+    /// Use `generator` to distinguish variables that appear in more than one positions.
+    pub fn set_equality_generator(&mut self, generator: Generator) {
+        self.equality_generator = generator;
+    }
+
+    fn helper(&self, formula: &Formula, vars: &mut HashMap<V, i32>) -> Result<Formula, Error> {
+        match formula {
+            Formula::Top | Formula::Bottom => Ok(formula.clone()),
+            Formula::Atom { predicate, terms } => {
+                let mut equations = Vec::new();
+                let mut new_terms = Vec::<Term>::new();
+                for term in terms {
+                    match term {
+                        Term::Var { variable } => {
+                            vars.entry(variable.clone())
+                                .and_modify(|count| {
+                                    let new_var = V(self
+                                        .equality_generator
+                                        .indexed_symbol(variable.to_string(), *count));
+                                    let eq_pred = Pred(self.equality_symbol.clone());
+                                    equations.push(eq_pred.app(vec![
+                                        variable.clone().into(),
+                                        new_var.clone().into(),
+                                    ]));
+                                    new_terms.push(new_var.into());
+                                    *count += 1;
+                                })
+                                .or_insert_with(|| {
+                                    new_terms.push(Term::Var {
+                                        variable: variable.clone(),
+                                    });
+                                    0
+                                });
+                        }
+                        _ => return Err(Error::EqualityExpandNonVar { term: term.clone() }),
```

Comment:
use `term_var!` instead (since the input is `Relational`). Also, because the input is `Relational`, this method cannot fail.

Reply:
Not really. The method already returns a `Result` + the caller may provide a bad input.